internal/http_cache/ghacache/ghacache.go

```diff
@@ -0,0 +1,221 @@
+			keys, version)
+		writer.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	writer.WriteHeader(http.StatusNoContent)
+}
+
+func (cache *GHACache) reserveUploadable(writer http.ResponseWriter, request *http.Request) {
+	var jsonReq struct {
+		Key     string `json:"key"`
+		Version string `json:"version"`
+	}
+
+	if err := render.DecodeJSON(request.Body, &jsonReq); err != nil {
+		log.Printf("GHA cache failed to read/decode the JSON passed to the "+
+			"reserve uploadable endpoint: %v\n", err)
+		writer.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	jsonResp := struct {
+		CacheID int64 `json:"cacheId"`
+	}{
+		CacheID: rand.Int63(),
+	}
+
+	cache.uploadables.Store(jsonResp.CacheID, &uploadable{
+		Key:     jsonReq.Key,
+		Version: jsonReq.Version,
+		Buf:     &bytes.Buffer{},
+	})
+
+	render.JSON(writer, request, &jsonResp)
+}
+
+func (cache *GHACache) updateUploadable(writer http.ResponseWriter, request *http.Request) {
+	id, ok := getID(request)
+	if !ok {
+		log.Printf("GHA cache failed to get/decode the ID passed to the " +
+			"update uploadable endpoint\n")
+		writer.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	uploadable, ok := cache.uploadables.Load(id)
+	if !ok {
+		writer.WriteHeader(http.StatusNotFound)
+
+		return
+	}
+
+	if _, err := io.Copy(uploadable.Buf, request.Body); err != nil {
+		if errors.Is(err, io.EOF) {
+			return
+		}
+
+		writer.WriteHeader(http.StatusInternalServerError)
+	}
+}
+
+func (cache *GHACache) commitUploadable(writer http.ResponseWriter, request *http.Request) {
+	id, ok := getID(request)
+	if !ok {
+		log.Printf("GHA cache failed to get/decode the ID passed to the " +
+			"commit uploadable endpoint\n")
+		writer.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	uploadable, ok := cache.uploadables.Load(id)
+	if !ok {
+		writer.WriteHeader(http.StatusNotFound)
+
+		return
+	}
+
+	var jsonReq struct {
+		Size int64 `json:"size"`
+	}
+
+	if err := render.DecodeJSON(request.Body, &jsonReq); err != nil {
+		log.Printf("GHA cache failed to read/decode the JSON passed to the "+
+			"commit uploadable endpoint: %v\n", err)
+		writer.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	if jsonReq.Size != int64(uploadable.Buf.Len()) {
+		log.Printf("GHA cache detected a cache entry size mismatch for uploadable "+
+			"with ID %d\n", id)
+		writer.WriteHeader(http.StatusBadRequest)
+
+		return
+	}
+
+	resp, err := http.Post(
+		cache.httpCacheURL(uploadable.Key, uploadable.Version),
+		"application/octet-stream",
+		bytes.NewReader(uploadable.Buf.Bytes()),
+	)
+	if err != nil {
+		log.Printf("GHA cache failed to upload the uploadable with ID %d: %v\n", id, err)
+		writer.WriteHeader(http.StatusInternalServerError)
+
+		return
+	}
+
+	if resp.StatusCode != http.StatusCreated {
+		log.Printf("GHA cache failed to upload the uploadable with ID %d: got HTTP %d\n",
+			id, resp.StatusCode)
+		writer.WriteHeader(http.StatusInternalServerError)
+
+		return
+	}
+
+	cache.uploadables.Delete(id)
```

Comment:
Should we `defer` it right after getting the `id`? Not sure if we need it in case of an error

Reply:
Good catch, fixed in https://github.com/cirruslabs/cirrus-ci-agent/pull/349/commits/6fc098f715cda97c96789a922f04e0c46211a4aa.